include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1988 @@
+                if(other.size_ > buffer_size)
+                {
+                    // self uses the buffer, other allocated memory
+                    for(index_t k=0; k<size_; ++k)
+                    {
+                        other.buffer_[k] = data_[k];
+                    }
+                    data_ = other.data_;
+                    other.data_ = other.buffer_;
+                }
+                else
+                {
+                    // both use the buffer
+                    if(size_ < other.size_)
+                    {
+                        for(index_t k=0; k<size_; ++k)
+                        {
+                            swap(data_[k], other.data_[k]);
+                        }
+                        for(index_t k=size_; k<other.size_; ++k)
+                        {
+                            data_[k] = other.data_[k];
+                        }
+                    }
+                    else
+                    {
+                        for(index_t k=0; k<other.size_; ++k)
+                        {
+                            swap(data_[k], other.data_[k]);
+                        }
+                        for(index_t k=other.size_; k<size_; ++k)
+                        {
+                            other.data_[k] = data_[k];
+                        }
+                    }
+                }
+            }
+            swap(size_, other.size_);
+        }
+
+        index_t size() const { return size_; }
+        index_t max_size() const { return std::max_size(alloc_); }
+        index_type shape() const { return { size_ }; }
+
+      protected:
+        void allocate(value_type const & initial = value_type())
+        {
+            if(size_ > buffer_size)
+            {
+                data_ = alloc_.allocate(size_);
+                std::uninitialized_fill(data_, data_+size_, initial);
+            }
+            else
+            {
+                data_ = buffer_;
+                std::fill(data_, data_+size_, initial);
+            }
+        }
+
+        void deallocate()
+        {
+            if(size_ > buffer_size)
+            {
+                if(!may_use_uninitialized_memory)
+                {
+                    for(index_t k=0; k<size_; ++k)
+                        (data_+k)->~value_type();
+                }
+                alloc_.deallocate(data_, size_);
+            }
+            size_ = 0;
+            data_ = nullptr;
+        }
+
+        template <class ITERATOR,
+                  XTENSOR_REQUIRE<iterator_concept<ITERATOR>::value>>
+        void init_impl(ITERATOR u)
+        {
+            for(index_t k=0; k < size_; ++k, ++u)
+            {
+                data_[k] = static_cast<value_type>(*u);
+            }
+        }
+
+        static const bool may_use_uninitialized_memory = std::is_scalar<value_type>::value ||
+                                                         std::is_pod<value_type>::value;
+        static const index_t buffer_size = 4;
```

Comment:
Why is this value hardcoded here?

Reply:
Because it is unfinished code...

However, looking at the implementation of the class, I came to the conclusion that an implementation in terms of a custom allocator (e.g. boost.pool) might be preferable. What do you think